zcash_client_sqlite/src/wallet/init/migrations/sapling_memo_consistency.rs

```diff
@@ -0,0 +1,215 @@
+//! ensure that memo entries are consistent with the decrypted transaction's outputs. The empty
+//! memo is now consistently represented as a single `0xf6` byte.
+
+use std::collections::{BTreeMap, HashMap, HashSet};
+
+use rusqlite::named_params;
+use schemer_rusqlite::RusqliteMigration;
+use uuid::Uuid;
+use zcash_client_backend::{decrypt_transaction, keys::UnifiedFullViewingKey};
+use zcash_primitives::{consensus, zip32::AccountId};
+
+use crate::{
+    error::SqliteClientError,
+    wallet::{get_transaction, init::WalletMigrationError, memo_repr},
+};
+
+use super::received_notes_nullable_nf;
+
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0x7029b904_6557_4aa1_9da5_6904b65d2ba5);
+
+pub(super) struct Migration<P> {
+    pub(super) params: P,
+}
+
+impl<P> schemer::Migration for Migration<P> {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        [received_notes_nullable_nf::MIGRATION_ID]
+            .into_iter()
+            .collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "This migration reads the wallet's raw transaction data and updates the `sent_notes` table to
+        ensure that memo entries are consistent with the decrypted transaction's outputs. The empty
+        memo is now consistently represented as a single `0xf6` byte."
+    }
+}
+
+impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction) -> Result<(), Self::Error> {
+        let mut stmt_raw_tx = transaction.prepare(
+            "SELECT DISTINCT sent_notes.tx, sent_notes.output_index, accounts.account, accounts.ufvk
+             FROM sent_notes 
+             JOIN accounts ON sent_notes.from_account = accounts.account
+             JOIN transactions ON transactions.id_tx = sent_notes.tx
+             WHERE transactions.raw IS NOT NULL"
+        )?;
+
+        let mut rows = stmt_raw_tx.query([])?;
+
+        let mut tx_sent_notes: BTreeMap<i64, Vec<(u32, AccountId, UnifiedFullViewingKey)>> =
+            BTreeMap::new();
+        while let Some(row) = rows.next()? {
+            let id_tx: i64 = row.get(0)?;
+            let output_index: u32 = row.get(1)?;
+            let account: u32 = row.get(2)?;
+            let ufvk_str: String = row.get(3)?;
+            let ufvk = UnifiedFullViewingKey::decode(&self.params, &ufvk_str).map_err(|e| {
+                WalletMigrationError::CorruptedData(format!(
+                    "Could not decode unified full viewing key for account {}: {:?}",
+                    account, e
+                ))
+            })?;
+
+            tx_sent_notes
+                .entry(id_tx)
+                .and_modify(|v| v.push((output_index, AccountId::from(account), ufvk.clone())))
+                .or_insert_with(|| vec![(output_index, AccountId::from(account), ufvk)]);
```

Comment:
Huh, I thought I left a comment here in my previous review, but I must have forgotten to click "Add". The comment was that we can avoid the `ufvk.clone()` here; merging that with the `output_index` removal we get:
```suggestion
            tx_sent_notes
                .entry(id_tx)
                .or_default()
                .insert(AccountId::from(account), ufvk);
```